WebKit::DataDetectionResults should re-use DDScannerResult instead of passing across an opaque NSArray.

gavin-apple opened the pull request:
#### 72ab7791cf7bfd84dad2e4387c75c4a31c56a1f1
<pre>
WebKit::DataDetectionResults should re-use DDScannerResult instead of passing across an opaque NSArray.
<a href="https://bugs.webkit.org/show_bug.cgi?id=301628">https://bugs.webkit.org/show_bug.cgi?id=301628</a>
<a href="https://rdar.apple.com/163644363">rdar://163644363</a>

Reviewed by Per Arne Vollan.

Replaces the IPC implementation of WebKit::DataDetectionResults to instead
serialize and explicit Vector of DDScannerResults.

* Source/WebKit/Shared/Cocoa/DataDetectionResult.h:
* Source/WebKit/Shared/Cocoa/DataDetectionResult.mm: Copied from Source/WebKit/Shared/Cocoa/DataDetectionResult.h.
(WebKit::DataDetectionResult::setResults):
* Source/WebKit/Shared/Cocoa/DataDetectionResult.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::setDataDetectionResult):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setDataDetectionResults):
(WebKit::detectDataInFrame):

Canonical link: <a href="https://commits.webkit.org/302560@main">https://commits.webkit.org/302560@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4e0007565d55ce6b953ed5d95e1523db6a436b1e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129432 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1689 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40271 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136809 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80849 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/62e1e7e8-bf49-40de-9e68-c4efd438ba5f) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1620 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1565 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98577 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66472 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ed657281-ff5d-4470-b88f-239247e26035) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132379 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1264 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115929 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79229 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/1376962d-841b-4541-91aa-b31239a8b45c) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1186 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80086 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109644 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34609 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139285 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1479 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1422 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107104 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1521 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112272 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106948 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27242 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1208 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30792 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54151 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1550 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64913 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1368 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1404 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1472 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

pvollan commented on Source/WebKit/Shared/Cocoa/DataDetectionResult.h:
```diff
@@ -32,11 +32,14 @@
 struct DataDetectionResult {
     WTF_DEPRECATED_MAKE_STRUCT_FAST_ALLOCATED(DataDetectionResult);
-    RetainPtr<NSArray> results;
+    void setResults(NSArray* detectionResults);
```
Nit: `NSArray*` -> `NSArray *`.

pvollan commented on Source/WebKit/Shared/Cocoa/DataDetectionResult.mm:
```diff
@@ -0,0 +1,51 @@
+
+namespace WebKit {
+
+void DataDetectionResult::setResults(NSArray* detectionResults)
```
Nit: `NSArray *`

pvollan commented on Source/WebKit/Shared/Cocoa/DataDetectionResult.mm:
```diff
@@ -0,0 +1,51 @@
+{
+    if ([arrayElement isKindOfClass:PAL::getDDScannerResultClassSingleton()])
+        return RetainPtr<DDScannerResult> { (DDScannerResult *)arrayElement };
+    return std::nullopt;
```
Can `checked_objc_cast` be used here?

pvollan commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -225,7 +225,9 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
 void WebPageProxy::setDataDetectionResult(DataDetectionResult&& dataDetectionResult)
 {
-    m_dataDetectionResults = WTFMove(dataDetectionResult.results);
+    m_dataDetectionResults = createNSArray(dataDetectionResult.results, [](const RetainPtr<DDScannerResult>& result) {
```
This may be outside the scope of this patch, but can `m_dataDetectionResults` also be of type  `Vector<RetainPtr<DDScannerResult>>`?

gavin-apple committed:
WebKit::DataDetectionResults should re-use DDScannerResult instead of passing across an opaque NSArray.
https://bugs.webkit.org/show_bug.cgi?id=301628
rdar://163644363

Reviewed by Per Arne Vollan.

Replaces the IPC implementation of WebKit::DataDetectionResults to instead
serialize and explicit Vector of DDScannerResults.

* Source/WebKit/Shared/Cocoa/DataDetectionResult.h:
* Source/WebKit/Shared/Cocoa/DataDetectionResult.mm: Copied from Source/WebKit/Shared/Cocoa/DataDetectionResult.h.
(WebKit::DataDetectionResult::setResults):
* Source/WebKit/Shared/Cocoa/DataDetectionResult.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::setDataDetectionResult):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setDataDetectionResults):
(WebKit::detectDataInFrame):

Canonical link: https://commits.webkit.org/302560@main
